testsuite/forge/src/interface/system_metrics.rs

```diff
@@ -0,0 +1,195 @@
+    if breach_pct > threshold.max_breach_pct {
+        bail!(
+            "Metrics violated threshold, max_breach_pct {:?}, breach_pct{:?} ",
+            threshold.max_breach_pct,
+            breach_pct
+        );
+    }
+    Ok(())
+}
+
+async fn query_prometheus_range_metrics(
+    query: &str,
+    client: &PrometheusClient,
+    start_time: i64,
+    end_time: i64,
+    internal_secs: f64,
+    namespace: &str,
+) -> anyhow::Result<Vec<Sample>> {
+    let mut labels_map = BTreeMap::new();
+    labels_map.insert("namespace".to_string(), namespace.to_string());
+    let response = client
+        .query_range(
+            construct_query_with_extra_labels(query, labels_map),
+            start_time,
+            end_time,
+            internal_secs,
+            None,
+        )
+        .await?;
+    Ok(response.as_range().unwrap()[0].samples().to_vec())
+}
+
+pub async fn query_prometheus_system_metrics(
+    client: &PrometheusClient,
+    start_time: i64,
+    end_time: i64,
+    internal_secs: f64,
+    namespace: &str,
+) -> anyhow::Result<SystemMetrics> {
+    let cpu_query = r#"avg(rate(container_cpu_usage_seconds_total{container=~"validator"}[30s]))"#;
+    let memory_query = r#"avg(container_memory_working_set_bytes{container=~"validator"})"#;
+
+    let cpu_samples = query_prometheus_range_metrics(
+        cpu_query,
+        client,
+        start_time,
+        end_time,
+        internal_secs,
+        namespace,
+    )
+    .await?;
+
+    let memory_samples = query_prometheus_range_metrics(
+        memory_query,
+        client,
+        start_time,
+        end_time,
+        internal_secs,
+        namespace,
+    )
+    .await?;
+
+    Ok(SystemMetrics::new(cpu_samples, memory_samples))
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::prometheus::get_prometheus_client;
+    use prometheus_http_query::Error as PrometheusError;
+
+    use super::*;
+    #[tokio::test]
+    async fn test_empty_metrics_threshold() {
+        let cpu_threshold = MetricsThreshold::new(10, 30);
+        let memory_threshold = MetricsThreshold::new(100, 40);
+        let threshold = SystemMetricsThreshold::new(cpu_threshold, memory_threshold);
+        let metrics = SystemMetrics::new(vec![], vec![]);
+        threshold.ensure_threshold(&metrics).unwrap_err();
+    }
+
+    #[tokio::test]
+    async fn test_query_prometheus() {
+        let client = get_prometheus_client().unwrap();
+        let query = r#"rate(container_cpu_usage_seconds_total{pod=~".*validator.*", container="validator"}[1m])"#;
+        let response = client.query(query, None, None).await;
+        match response {
+            Ok(pres) => println!("{:?}", pres),
+            Err(PrometheusError::Client(e)) => {
+                println!("Skipping test. Failed to create prometheus client: {}", e);
+                return;
+            }
+            Err(e) => panic!("Expected PromqlResult: {}", e),
+        }
+
+        let start_timestamp = 1660453807;
```

Comment:
will this test always run? seems like we're hitting the actual prometheus server, and the timestamp might get stale. We might want to get rid of this test or mock the prometheus server

Reply:
Yes this was accidentally added - I used this for my local testing.